File: sdk-java/src/test/java/io/superdurable/dex/integ/stateoptions/StateOptionsWorkflow.java
/*
 * Legacy Materials in this file remain under their original licenses.
 * See LEGACY_NOTICES.md.
 */

/*
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications after the Legacy Cutoff are licensed under the
 * Super Durable Source License 1.0.
 * Legacy Materials remain under their original licenses.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ.stateoptions;

import io.superdurable.dex.core.Context;
import io.superdurable.dex.core.ObjectWorkflow;
import io.superdurable.dex.core.StateDecision;
import io.superdurable.dex.core.StateDef;
import io.superdurable.dex.core.WorkflowState;
import io.superdurable.dex.core.WorkflowStateOptions;
import io.superdurable.dex.core.command.CommandRequest;
import io.superdurable.dex.core.command.CommandResults;
import io.superdurable.dex.core.communication.Communication;
import io.superdurable.dex.core.persistence.DataAttributeDef;
import io.superdurable.dex.core.persistence.Persistence;
import io.superdurable.dex.core.persistence.PersistenceFieldDef;
import io.superdurable.dex.gen.models.PersistenceLoadingPolicy;
import io.superdurable.dex.gen.models.PersistenceLoadingType;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@Component
public class StateOptionsWorkflow implements ObjectWorkflow {
    public static final String DA_WAIT_UNTIL = "DA_WAIT_UNTIL";
    public static final String DA_EXECUTE = "DA_EXECUTE";
    public static final String DA_BOTH = "DA_BOTH";

    @Override
    public List<PersistenceFieldDef> getPersistenceSchema() {
        return Arrays.asList(
                DataAttributeDef.create(String.class, DA_WAIT_UNTIL),
                DataAttributeDef.create(String.class, DA_EXECUTE),
                DataAttributeDef.create(String.class, DA_BOTH)
        );
    }

    @Override
    public List<StateDef> getWorkflowStates() {
        return Arrays.asList(
                StateDef.startingState(new StateOptionsWorkflowState1()),
                StateDef.nonStartingState(new StateOptionsWorkflowState2()),
                StateDef.nonStartingState(new StateOptionsWorkflowState3())
        );
    }
}

class StateOptionsWorkflowState1 implements WorkflowState<Void> {
    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest waitUntil(final Context context, final Void input, Persistence persistence, final Communication communication) {
        return CommandRequest.empty;
    }

    @Override
    public StateDecision execute(final Context context, final Void input, final CommandResults commandResults, Persistence persistence, final Communication communication) {
        persistence.setDataAttribute(StateOptionsWorkflow.DA_EXECUTE, "execute");
        persistence.setDataAttribute(StateOptionsWorkflow.DA_WAIT_UNTIL, "wait_until");
        persistence.setDataAttribute(StateOptionsWorkflow.DA_BOTH, "both");

        return StateDecision.singleNextState(StateOptionsWorkflowState2.class);
    }
}

class StateOptionsWorkflowState2 implements WorkflowState<Void> {
    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest waitUntil(final Context context, final Void input, Persistence persistence, final Communication communication) {
        final String daWaitUntil = persistence.getDataAttribute(StateOptionsWorkflow.DA_WAIT_UNTIL, String.class);

        if (!daWaitUntil.equals("wait_until")) {
            throw new RuntimeException("Expected DA_WAIT_UNTIL to be 'wait_until', got " + daWaitUntil);
        }

        final String daExecute = persistence.getDataAttribute(StateOptionsWorkflow.DA_EXECUTE, String.class);
        if (daExecute != null) {
            throw new RuntimeException("Expected DA_EXECUTE to be null, got " + daExecute);
        }

        return CommandRequest.empty;
    }

    @Override
    public StateDecision execute(final Context context, final Void input, final CommandResults commandResults, Persistence persistence, final Communication communication) {
        final String daExecute = persistence.getDataAttribute(StateOptionsWorkflow.DA_EXECUTE, String.class);

        if (!daExecute.equals("execute")) {
            throw new RuntimeException("Expected DA_EXECUTE to be 'execute', got " + daExecute);
        }

        final String daWaitUntil = persistence.getDataAttribute(StateOptionsWorkflow.DA_WAIT_UNTIL, String.class);
        if (daWaitUntil != null) {
            throw new RuntimeException("Expected DA_WAIT_UNTIL to be null, got " + daWaitUntil);
        }

        return StateDecision.singleNextState(StateOptionsWorkflowState3.class);
    }

    @Override
    public WorkflowStateOptions getStateOptions() {
        return new WorkflowStateOptions()
                .setWaitUntilApiDataAttributesLoadingPolicy(
                        new PersistenceLoadingPolicy()
                                .persistenceLoadingType(PersistenceLoadingType.PARTIAL_WITH_EXCLUSIVE_LOCK)
                                .partialLoadingKeys(Collections.singletonList(StateOptionsWorkflow.DA_WAIT_UNTIL))
                )
                .setExecuteApiDataAttributesLoadingPolicy(
                        new PersistenceLoadingPolicy()
                                .persistenceLoadingType(PersistenceLoadingType.PARTIAL_WITH_EXCLUSIVE_LOCK)
                                .partialLoadingKeys(Collections.singletonList(StateOptionsWorkflow.DA_EXECUTE)));
    }
}

class StateOptionsWorkflowState3 implements WorkflowState<Void> {
    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest waitUntil(final Context context, final Void input, Persistence persistence, final Communication communication) {
        final String str = persistence.getDataAttribute(StateOptionsWorkflow.DA_BOTH, String.class);

        if (!str.equals("both")) {
            throw new RuntimeException("Expected DA_BOTH to be 'both', got " + str);
        }

        return CommandRequest.empty;
    }

    @Override
    public StateDecision execute(final Context context, final Void input, final CommandResults commandResults, Persistence persistence, final Communication communication) {
        final String str = persistence.getDataAttribute(StateOptionsWorkflow.DA_BOTH, String.class);

        if (!str.equals("both")) {
            throw new RuntimeException("Expected DA_BOTH to be 'both', got " + str);
        }

        return StateDecision.gracefulCompleteWorkflow("success");
    }

    @Override
    public WorkflowStateOptions getStateOptions() {
        return new WorkflowStateOptions()
                .setDataAttributesLoadingPolicy(
                        new PersistenceLoadingPolicy()
                                .persistenceLoadingType(PersistenceLoadingType.PARTIAL_WITH_EXCLUSIVE_LOCK)
                                .partialLoadingKeys(Collections.singletonList(StateOptionsWorkflow.DA_BOTH))
                );
    }
}

File: sdk-python/dex/tests/workflows/state_options_workflow.py
# Legacy Materials in this file remain under their original licenses.
# See LEGACY_NOTICES.md.

# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications after the Legacy Cutoff are licensed under the
# Super Durable Source License 1.0.
# Legacy Materials remain under their original licenses.
# See LICENSE and LEGACY_NOTICES.md.

from dex.command_request import CommandRequest
from dex.command_results import CommandResults
from dex.communication import Communication

from dex.dex_api.models import RetryPolicy, WaitUntilApiFailurePolicy
from dex.persistence import Persistence
from dex.state_decision import StateDecision
from dex.state_schema import StateSchema
from dex.workflow import ObjectWorkflow
from dex.workflow_context import WorkflowContext
from dex.workflow_state import T, WorkflowState
from dex.workflow_state_options import WorkflowStateOptions

class InitState1(WorkflowState[str]):
    def get_state_options(self) -> WorkflowStateOptions:
        return WorkflowStateOptions(
            wait_until_api_retry_policy=RetryPolicy(maximum_attempts=1),
            proceed_to_execute_when_wait_until_retry_exhausted=WaitUntilApiFailurePolicy.PROCEED_ON_FAILURE,
        )

    def wait_until(
        self,
        ctx: WorkflowContext,
        input: T,
        persistence: Persistence,
        communication: Communication,
    ) -> CommandRequest:
        raise RuntimeError("test failure")

    def execute(
        self,
        ctx: WorkflowContext,
        input: T,
        command_results: CommandResults,
        persistence: Persistence,
        communication: Communication,
    ) -> StateDecision:
        data = "InitState1_execute_completed"
        return StateDecision.graceful_complete_workflow(data)

class InitState2(WorkflowState[str]):
    def get_state_options(self) -> WorkflowStateOptions:
        return WorkflowStateOptions(
            wait_until_api_retry_policy=RetryPolicy(maximum_attempts=1),
            proceed_to_execute_when_wait_until_retry_exhausted=WaitUntilApiFailurePolicy.FAIL_WORKFLOW_ON_FAILURE,
        )

    def wait_until(
        self,
        ctx: WorkflowContext,
        input: T,
        persistence: Persistence,
        communication: Communication,
    ) -> CommandRequest:
        raise RuntimeError("test failure")

    def execute(
        self,
        ctx: WorkflowContext,
        input: T,
        command_results: CommandResults,
        persistence: Persistence,
        communication: Communication,
    ) -> StateDecision:
        data = "InitState2_execute_completed"
        return StateDecision.graceful_complete_workflow(data)

class StateOptionsWorkflow1(ObjectWorkflow):
    def get_workflow_states(self) -> StateSchema:
        return StateSchema.with_starting_state(InitState1())

class StateOptionsWorkflow2(ObjectWorkflow):
    def get_workflow_states(self) -> StateSchema:
        return StateSchema.with_starting_state(InitState2())
